// annotator/annotator.debugstore.js
/*
** Annotator v2.0.0-dev-41928d6
** https://github.com/okfn/annotator/
**
** Copyright 2014, the Annotator project contributors.
** Dual licensed under the MIT and GPLv3 licenses.
** https://github.com/okfn/annotator/blob/master/LICENSE
**
** Built at: 2014-04-11 02:53:35Z
*/
!function(e){if("object"==typeof exports)module.exports=e();else if("function"==typeof define&&define.amd)define(e);else{var n;"undefined"!=typeof window?n=window:"undefined"!=typeof global?n=global:"undefined"!=typeof self&&(n=self);var o=n;o=o.Annotator||(o.Annotator={}),o=o.Plugin||(o.Plugin={}),o.Debugstore=e()}}(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);throw new Error("Cannot find module '"+o+"'")}var f=n[o]={exports:{}};t[o][0].call(f.exports,function(e){var n=t[o][1][e];return s(n?n:e)},f,f.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({"4hR4NO":[function(_dereq_,module,exports){
(function (global){
var Annotator, self, _ref;

if (typeof self !== "undefined" && self !== null) {
  self = self;
}

if (typeof global !== "undefined" && global !== null) {
  if (self == null) {
    self = global;
  }
}

if (typeof window !== "undefined" && window !== null) {
  if (self == null) {
    self = window;
  }
}

Annotator = self != null ? self.Annotator : void 0;

if (Annotator == null) {
  Annotator = (self != null ? (_ref = self.define) != null ? _ref.amd : void 0 : void 0) ? self != null ? self.require('annotator') : void 0 : void 0;
}

if (typeof Annotator !== 'function') {
  throw new Error("Could not find Annotator! In a webpage context, please ensure that the Annotator script tag is loaded before any plugins.");
}

module.exports = Annotator;


}).call(this,typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{}],"annotator":[function(_dereq_,module,exports){
module.exports=_dereq_('4hR4NO');
},{}],3:[function(_dereq_,module,exports){
var $, Annotator, DebugStore, log, perform, uuid,
  __slice = [].slice;

Annotator = _dereq_('annotator');

$ = Annotator.Util.$;

uuid = (function() {
  var counter;
  counter = -1;
  return function() {
    return counter += 1;
  };
})();

log = function() {
  var args;
  args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
  return console.debug.apply(console, ["DebugStore"].concat(__slice.call(args)));
};

perform = function(action, annotation) {
  var dfd;
  log(action, JSON.parse(JSON.stringify(annotation)));
  dfd = $.Deferred();
  dfd.resolve(annotation);
  return dfd.promise();
};

DebugStore = (function() {
  function DebugStore() {}

  DebugStore.prototype.create = function(annotation) {
    annotation.id = uuid();
    return perform('create', annotation);
  };

  DebugStore.prototype.update = function(annotation) {
    return perform('update', annotation);
  };

  DebugStore.prototype["delete"] = function(annotation) {
    return perform('destroy', annotation);
  };

  DebugStore.prototype.query = function(queryObj) {
    var dfd;
    dfd = $.Deferred();
    perform('query', queryObj);
    dfd.resolve([], {
      total: 0
    });
    return dfd.promise();
  };

  DebugStore.prototype.setHeader = function(key, value) {
    return log("would set header '" + key + "'='" + value + "'");
  };

  return DebugStore;

})();

Annotator.Plugin.register('DebugStore', DebugStore);

module.exports = DebugStore;


},{}]},{},[3])

(3)
});
